Returning a warning in ValidateFunc will cause a panic in apply step

When adding a `ValidateFunc` for any schema, returning a non-nil `warnings` slice will cause the a provider to _not_ call it's configure block. 

Branch [b-validate-func-config][1] demonstrates this: 

- https://github.com/hashicorp/terraform/commit/5e877fe161d68ab8ef2f68e9080156dc4b36b38c

`plan` will show the error:

```
Warnings:

  * provider.aws: this is a warning

No errors found. Continuing with 1 warning(s).

Refreshing Terraform state prior to plan...

```

apply causes a panic:

```
2015/07/28 11:10:05 [DEBUG] apply: aws_instance.nat: executing Apply
2015/07/28 11:10:05 terraform-provider-aws: aws_instance.nat: Creating...
  ami:                      "" => "ami-21f78e11"
  availability_zone:        "" => "<computed>"
  ebs_block_device.#:       "" => "<computed>"
  ephemeral_block_device.#: "" => "<computed>"
  instance_type:            "" => "t1.micro"
  key_name:                 "" => "<computed>"
  placement_group:          "" => "<computed>"
  private_dns:              "" => "<computed>"
  private_ip:               "" => "<computed>"
  public_dns:               "" => "<computed>"
  public_ip:                "" => "<computed>"
  root_block_device.#:      "" => "<computed>"
  security_groups.#:        "" => "<computed>"
  source_dest_check:        "" => "1"
  subnet_id:                "" => "<computed>"
  tags.#:                   "" => "1"
  tags.Name:                "" => "testing-vpc-subnet-things"
  tenancy:                  "" => "<computed>"
  vpc_security_group_ids.#: "" => "<computed>"
Error applying plan:

2 error(s) occurred:

* aws_vpc.foo: unexpected EOF
* aws_instance.nat: unexpected EOF

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
panic: interface conversion: interface is nil, not *aws.AWSClient
2015/07/28 11:10:05 terraform-provider-aws:
2015/07/28 11:10:05 terraform-provider-aws: goroutine 26 [running]:
2015/07/28 11:10:05 terraform-provider-aws: github.com/hashicorp/terraform/builtin/providers/aws.resourceAwsVpcCreate(0xc2080520c0, 0x0, 0x0, 0x0, 0x0)
2015/07/28 11:10:05 terraform-provider-aws: 	/Users/clint/Projects/Go/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_vpc.go:84 +0x67
2015/07/28 11:10:05 terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc208142c80, 0xc2081829c0, 0xc20817d5c0, 0x0, 0x0, 0x1, 0x0, 0x0)
2015/07/28 11:10:05 terraform-provider-aws: 	/Users/clint/Projects/Go/src/github.com/hashicorp/terraform/helper/schema/resource.go:145 +0x2c1
2015/07/28 11:10:05 terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc2080b5620, 0xc208185480, 0xc2081829c0, 0xc20817d5c0, 0x6b2920, 0x0, 0x0)
2015/07/28 11:10:05 terraform-provider-aws: 	/Users/clint/Projects/Go/src/github.com/hashicorp/terraform/helper/schema/provider.go:162 +0x1da
2015/07/28 11:10:05 terraform-provider-aws: github.com/hashicorp/terraform/rpc.(*ResourceProviderServer).Apply(0xc2080ed700, 0xc20817b940, 0xc20817d6a0, 0x0, 0x0)
2015/07/28 11:10:05 terraform-provider-aws: 	/Users/clint/Projects/Go/src/github.com/hashicorp/terraform/rpc/resource_provider.go:323 +0x7f
2015/07/28 11:10:05 terraform-provider-aws: reflect.Value.call(0x866040, 0xa47f20, 0x13, 0xab48a0, 0x4, 0xc20809bf28, 0x3, 0x3, 0x0, 0x0, ...)
2015/07/28 11:10:05 terraform-provider-aws: 	/usr/local/go-versions/go1.4.2/src/reflect/value.go:419 +0x10e5
2015/07/28 11:10:05 terraform-provider-aws: reflect.Value.Call(0x866040, 0xa47f20, 0x13, 0xc20809bf28, 0x3, 0x3, 0x0, 0x0, 0x0)
2015/07/28 11:10:05 terraform-provider-aws: 	/usr/local/go-versions/go1.4.2/src/reflect/value.go:296 +0xbc
2015/07/28 11:10:05 terraform-provider-aws: net/rpc.(*service).call(0xc208142e80, 0xc208142e40, 0xc20800b8a8, 0xc2080a2880, 0xc208162000, 0x6b28c0, 0xc20817b940, 0x16, 0x6b2920, 0xc20817d6a0, ...)
2015/07/28 11:10:05 terraform-provider-aws: 	/usr/local/go-versions/go1.4.2/src/net/rpc/server.go:382 +0x1f7
2015/07/28 11:10:05 terraform-provider-aws: created by net/rpc.(*Server).ServeCodec
2015/07/28 11:10:05 terraform-provider-aws: 	/usr/local/go-versions/go1.4.2/src/net/rpc/server.go:476 +0x44a
2015/07/28 11:10:05 terraform-provider-aws:
```

cc @phinze / @mitchellh 



[1]: https://github.com/hashicorp/terraform/tree/b-validate-func-config

--- a/terraform/context_apply_test.go
+++ b/terraform/context_apply_test.go
@@ -78,6 +78,46 @@ func TestContext2Apply_providerAlias(t *testing.T) {
 	}
 }
 
+// GH-2870
+func TestContext2Apply_providerWarning(t *testing.T) {
+	m := testModule(t, "apply-provider-warning")
+	p := testProvider("aws")
+	p.ApplyFn = testApplyFn
+	p.DiffFn = testDiffFn
+	p.ValidateFn = func(c *ResourceConfig) (ws []string, es []error) {
+		ws = append(ws, "Just a warning")
+		return
+	}
+	ctx := testContext2(t, &ContextOpts{
+		Module: m,
+		Providers: map[string]ResourceProviderFactory{
+			"aws": testProviderFuncFixed(p),
+		},
+	})
+
+	if _, err := ctx.Plan(); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	state, err := ctx.Apply()
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	actual := strings.TrimSpace(state.String())
+	expected := strings.TrimSpace(`
+aws_instance.foo:
+  ID = foo
+	`)
+	if actual != expected {
+		t.Fatalf("got: \n%s\n\nexpected:\n%s", actual, expected)
+	}
+
+	if !p.ConfigureCalled {
+		t.Fatalf("provider Configure() was never called!")
+	}
+}
+
 func TestContext2Apply_emptyModule(t *testing.T) {
 	m := testModule(t, "apply-empty-module")
 	p := testProvider("aws")

--- a/terraform/evaltree_provider.go
+++ b/terraform/evaltree_provider.go
@@ -40,9 +40,8 @@ func ProviderEvalTree(n string, config *config.RawConfig) EvalNode {
 		},
 	})
 
-	// Apply stuff
 	seq = append(seq, &EvalOpFilter{
-		Ops: []walkOperation{walkValidate, walkRefresh, walkPlan, walkApply},
+		Ops: []walkOperation{walkValidate},
 		Node: &EvalSequence{
 			Nodes: []EvalNode{
 				&EvalGetProvider{
@@ -61,6 +60,32 @@ func ProviderEvalTree(n string, config *config.RawConfig) EvalNode {
 				&EvalValidateProvider{
 					Provider: &provider,
 					Config:   &resourceConfig,
+				},
+				&EvalSetProviderConfig{
+					Provider: n,
+					Config:   &resourceConfig,
+				},
+			},
+		},
+	})
+
+	// Apply stuff
+	seq = append(seq, &EvalOpFilter{
+		Ops: []walkOperation{walkRefresh, walkPlan, walkApply},
+		Node: &EvalSequence{
+			Nodes: []EvalNode{
+				&EvalGetProvider{
+					Name:   n,
+					Output: &provider,
+				},
+				&EvalInterpolate{
+					Config: config,
+					Output: &resourceConfig,
+				},
+				&EvalBuildProviderConfig{
+					Provider: n,
+					Config:   &resourceConfig,
+					Output:   &resourceConfig,
 				},
 				&EvalSetProviderConfig{
 					Provider: n,

--- a/terraform/test-fixtures/apply-provider-warning/main.tf
+++ b/terraform/test-fixtures/apply-provider-warning/main.tf
@@ -1,0 +1,1 @@
+resource "aws_instance" "foo" {}

